aten/src/ATen/native/UnaryOps.cpp

```diff
@@ -149,17 +149,24 @@ Tensor& deg2rad_out(Tensor& result, const Tensor& self) {
 Tensor deg2rad(const Tensor& self) { return unary_op_impl(self, at::deg2rad_out); }
 Tensor& deg2rad_(Tensor& self) { return unary_op_impl_(self, at::deg2rad_out); }
 
-Tensor& asin_out(Tensor& result, const Tensor& self) { return unary_op_impl_out(result, self, asin_stub); }
-Tensor asin(const Tensor& self) { return unary_op_impl(self, at::asin_out); }
+Tensor& asin_out(Tensor& result, const Tensor& self) { return unary_op_impl_float_out(result, self, asin_stub); }
+Tensor asin(const Tensor& self) {
+  if (self.is_sparse()) {
```

Comment:
Can you elaborate on what's going on here? In particular, how does sparse asin end up here? Maybe it should dispatch to something else?

Reply:
Looking at the entry below ->
https://github.com/pytorch/pytorch/blob/ee0033af9b7531ba0a453844dc621f8cfa140c20/aten/src/ATen/native/native_functions.yaml#L634-L650

Only inplace and out variant have a special dispatch for sparse.

So when we call `asin`, we allocate an output tensor and call the out variant.

TensorIterator doesn't work if input tensor is sparse and out is undefined tensor as it throws `RuntimeError: unsupported tensor layout: Sparse` (the path with `unary_op_impl_float`), 
so for sparse we use the previously working path `unary_op_impl` which defines the output tensor to have same dtype and layout as input.

But thinking about that now. It means sparse tensor wouldn't promote from integer to floats. :thinking: 
Will have to look for a better solution.
Thanks for pointing that out!